nixos/modules/services/backup/borgbackup.nix

```diff
@@ -0,0 +1,531 @@
+    borg create \
+      --compression ${cfg.compression} \
+      --exclude-from ${mkExcludeFile cfg} \
+      $extraCreateArgs \
+      "::$archiveName$archiveSuffix" \
+      ${escapeShellArgs cfg.paths}
+  '' + optionalString cfg.appendFailedSuffix ''
+    lastArchive=$(borg list --short | tail -1)
+    borg rename "::$lastArchive" ''${lastArchive%.failed}
+  '' + ''
+    ${cfg.postCreate}
+  '' + optionalString (cfg.prune.keep != { }) ''
+    borg prune \
+      ${mkKeepArgs cfg} \
+      $extraPruneArgs
+    ${cfg.postPrune}
+  '';
+
+  mkPassEnv = cfg: with cfg.encryption;
+    if passCommand != null then
+      { BORG_PASSCOMMAND = passCommand; }
+    else if passphrase != null then
+      { BORG_PASSPHRASE = passphrase; }
+    else { };
+
+  mkBackupService = name: cfg: 
+    let
+      userHome = config.users.users.${cfg.user}.home;
+    in nameValuePair "borgbackup-job-${name}" {
+      description = "BorgBackup job ${name}";
+      path = with pkgs; [
+        borgbackup openssh
+      ];
+      script = mkBackupScript cfg;
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        # Only run when no other process is using CPU or disk
+        CPUSchedulingPolicy = "idle";
+        IOSchedulingClass = "idle";
+        ProtectSystem = "strict";
+        ReadWritePaths =
+          [ "${userHome}/.config/borg" "${userHome}/.cache/borg" ]
+          # Borg needs write access to repo if it is not remote
+          ++ optional (isPath cfg.repo) cfg.repo;
+        PrivateTmp = true;
+      };
+      environment = {
+        BORG_REPO = cfg.repo;
+        inherit (cfg) archiveName;
+        inherit (cfg) extraInitArgs extraCreateArgs extraPruneArgs;
+      } // (mkPassEnv cfg) // cfg.environment;
+      inherit (cfg) startAt;
+    };
+
+  # The directories listed in ReadWritePaths need to be created
+  # before the service is started
+  mkActivationScript = name: cfg:
+    let
+      install = "install -o ${cfg.user} -g ${cfg.group}";
+    in
+      nameValuePair "borgbackup-job-${name}" (stringAfter [ "users" ] (
+        ''
+          cd "${config.users.users.${cfg.user}.home}"                                                                                                         
+          ${install} -d .config/borg
+          ${install} -d .cache/borg
+        '' + optionalString (isPath cfg.repo) ''                                                                                                              
+          ${install} -d ${escapeShellArg cfg.repo}
+        ''
+      ));
+
+  # The name arg is needed for mapAttrsToList
+  mkPassAssertion = name: cfg: {
+    assertion = with cfg.encryption;
+      mode != "none" -> passCommand != null || passphrase != null;
+    message =
+      ''passCommand or passphrase have to be specified ''
+      + ''because borgbackup.jobs.${name}.encryption != "none"'';
+  };
+
+  mkRepoService = name: cfg:
+    nameValuePair "borgbackup-repo-${name}" {
```

Comment:
borgbackup-repo -> borgbackup-create-repo ? (OCD, I know...)

Reply:
Au contraire. I highly appreciate your thorough review :)

Reply:
I think I take it back. I see how the current name maps nicely to the configuration name (borgbackup.repo).

Reply:
I'm wondering if I should just drop the repo service and use an activation script instead.

Reply:
I like the service approach better. It allows configuring more detailed dependency graph, and you'll be able to wipe directories and restart the service without a problem. (Using activation scripts means you'll have to reboot to setup the service again.)